### ISSUE TITLE
Where to implement SentencePiece subword tokenization?

### ISSUE THREAD

**Issue by sai-prasanna** [allenai/allennlp#1570]
I am new to allennlp, I use sentencepiece for subword tokenization in my pipeline.

SentencePiece needs a training step to generate a subword model, which can then be used for tokenization.

Is an implementation of Vocabulary class the right way to do it. Little confused whether it is the right place, given there are TokenIndexers etc.

And if sentencepiece dependency is not too specific for allennlp, I can give a pull request.

**Comment by schmmd** [comment]
@matt-peters @DeNeutoy do either of you have an idea here?

**Comment by sai-prasanna** [comment]
From further reading of the source code, I found that extending vocabulary might not be the right solution.
SentencePiece/Any BPE are tokenizers that can take a sentence and tokenize it. So they fit tokenizer part. But the problem is these tokenizers need a subword model to be trained on all the sentences before they can create the tokens. 

Should we add a training phase to Tokenizer class , so that this get addressed everywhere, or is there any other way to do this without changing allennlp library.

**Comment by joelgrus** [comment]
off the top of my head, here's probably the easiest way to do this:

(1) create a subword Model and train that (and get a `model.tar.gz` file out it)
(2) create a custom `Tokenizer` that takes that trained model as one of its constructor parameters, and train normally using that

that doesn't address subword regularization. as I understand it, you want to allow a different tokenization for the same sentence each epoch? in that case I might do the following:

(1) again, create and train the subword Model
(2) pass in the subword model as a constructor to your DatasetReader, make your dataset reader lazy, and do the tokenization inside `DatasetReader._read()`, then your instance will have to have a `ListField[TextField]` for your sentence instead of just a `TextField`, and each epoch you'll get a different tokenization. in this case you'll have to create the vocabulary ahead of time as well, as the "generate it based on a pass through the dataset" won't necessarily generate all possibilities.

let me know if that makes sense, maybe someone else has a better idea, but those are what I would try.

**Comment by sai-prasanna** [comment]
@joelgrus  I made a rough version for handling BPE training/tokenization.

https://gist.github.com/sai-prasanna/4749d84aa2d077d76d71058e54556e64

I have implemented the training part inside the SentencePieceTokenizer class. But this needed me to write a new in Dataset reader (here for Seq2Seq), since I would have to pass all the sentence to be trained.

I did most changes in `read` method. I have made an attempt to handle lazy workflow too.

``` python
@overrides
    def _read(self, file_path):
        
        source_trained, target_trained = True, True
        if isinstance(self._source_tokenizer, SentencePieceTokenizer):
            source_trained = self._source_tokenizer.trained
        if isinstance(self._target_tokenizer, SentencePieceTokenizer):
            target_trained = self._target_tokenizer.trained
        
        any_untrained = not (source_trained and target_trained)
        
        if self.lazy and any_untrained:
            raise ConfigurationError(
                "Cannot run lazily without both source_tokenizer and target_tokenizer trained")

        if any_untrained:
            sources, targets = [], []
        
        with open(cached_path(file_path), "r") as data_file:
            logger.info("Reading instances from lines in file at: %s", file_path)
            for line_num, line in enumerate(data_file):
                line = line.strip("\n")

                if not line:
                    continue

                line_parts = line.split('\t')
                if len(line_parts) != 2:
                    raise ConfigurationError("Invalid line format: %s (line number %d)" % (line, line_num + 1))
                
                source_sequence, target_sequence = line_parts
                if any_untrained:
                    sources.append(source_sequence)
                    targets.append(target_sequence)
                else:
                    yield self.text_to_instance(source_sequence, target_sequence)
                    
        if any_untrained:
            if self._source_tokenizer == self._target_tokenizer:
                self._source_tokenizer.train(sources+targets)
            else:
                self._source_tokenizer.train(sources)
                self._target_tokenizer.train(targets)
            for source_sequence, target_sequence in zip(sources, targets):
                yield self.text_to_instance(source_sequence, target_sequence) 
```
As you can see the ugly isinstanceof comes from the fact that current Tokenizer superclass doesn't have explicit training API.

Is adding train method, and trained flag a good path to go, If I were to implement it in generic way ?

**Comment by sai-prasanna** [comment]
I felt the lack of access to serialization_dir from Tokenizer impeding. Maybe we can think about an API where objects down the dependency graph can access some globals, which maybe ugly.

Or provide an serialization API, so that objects down the graph can serialize stuff. Im not sure whether this is too specific need, or can help for other stuff. One use case off my mind for a global serialization API is serializing processed sentences so that we don't compute them again.

**Comment by joelgrus** [comment]
what I was recommending was slightly different:

(1) write code to train your tokenizer (independent of the rest of your model) and run that training loop in the usual way (producing a `model.tar.gz` corresponding to the tokenizer)

(2) then write a dataset reader that takes that trained model as input and run a whole second training loop.

--

it seems like a lot of the complications you're running into are from trying to do it all in one pass, rather than in two passes.

**Comment by sai-prasanna** [comment]
In that approach, I would not be able to try configurations like effect of splitting (normalizing) words with spacy before sentencepiece etc without redundant code.

Even in the pretrain tokenizer before running allen nlp case wouldn't access to serialization (or simple injection of the directory) make sense. At least allowing tokenizer etc. to copy the model into the serialization directory and load from it?

**Comment by sai-prasanna** [comment]
Closing this, I think what @joelgrus  suggested is good solution. 
Will raise separate issue making case for a serialization API handle if its important.
